### PR TITLE
different types for different speedy phases

### PR DESCRIPTION
**Pull request by nickchapman-da**
Use different expression types for pre/post ANF transformation
- `SExpr` is the type executed on a speedy machine; clients of speedy use this type.
- `SExpr0` is the type used during the speedy compilation pipeline; the type is purely internal to the speedy compiler pipeline
- The ANF compilation phase transforms from `SExp0` to `SExpr`
- The two types contains only the variant forms required.

The intention is that this refactoring (and others) will help with the more substantive issue: https://github.com/digital-asset/daml/issues/11561, by reducing the cases we have to consider when making the various speedy compilation phases stack-safe

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import com.daml.lf.value.{Value => V}
+import com.daml.lf.speedy.SValue._
+import com.daml.lf.speedy.{SExpr => t}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+object SExpr0 {
+
+  sealed abstract class SExpr extends Product with Serializable
+
+  sealed abstract class SExprAtomic extends SExpr
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala:38]:
```suggestion

sealed abstract class SExpr0 extends Product with Serializable

object SExpr0 {
  
  sealed abstract class SEAtomic extends SExpr0
  
    /** Reference to a value. On first lookup the evaluated expression is
    * stored in 'cached'.
    */
  final case class SEVal(ref: t.SDefinitionRef) extends SExpr
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala [hunk 2]
```diff
@@ -46,7 +46,6 @@ private[speedy] object PrettyLightweight { // lightweight pretty printer for CEK
 
   def pp(e: SExpr): String = e match {
     case SEValue(v) => s"(VALUE)${pp(v)}"
-    case SEVar(n) => s"D#$n" //dont expect these at runtime
     case loc: SELoc => pp(loc)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala:49]: :partying_face:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -13,8 +13,10 @@ import com.daml.lf.interpretation.{Error => IE}
 import com.daml.lf.language.Ast
 import com.daml.lf.speedy.SError._
 import com.daml.lf.speedy.SExpr._
-import com.daml.lf.speedy.Speedy._
 import com.daml.lf.speedy.SResult._
+import com.daml.lf.speedy.Speedy._
+import com.daml.lf.speedy.{SExpr0 => s}
+import com.daml.lf.speedy.{SExpr => t}
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:19]: Do not rename here,
-  convention source/target does not make sens here 
- the file should not use SExpr0 (We can tackle the second point is an separate PR)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 4]
```diff
@@ -514,7 +460,9 @@ object SExpr {
       final case object EqualList extends Reference
     }
 
-    val EqualList: SEBuiltinRecursiveDefinition = SEBuiltinRecursiveDefinition(Reference.EqualList)
+    private val EqualList: SEBuiltinRecursiveDefinition = SEBuiltinRecursiveDefinition(
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:463]: :+1: 

```suggestion
    private[this] val EqualList: SEBuiltinRecursiveDefinition = SEBuiltinRecursiveDefinition(
```

**Review by remyhaemmerle-da — APPROVED** [review]
I like very much how everything is going. Very good work and thank a lot for pushing in this direction.

However, I do not like the convention to `source`/`target`.
Ideally we should have come with a good name for those two types, and not have to rename them at all.
On the one hand, `SExpr` for the type that the interpreter interprets is definitively the proper one, however I cannot come with a better one for `SExpr0` :cry: .

Assuming we cannot find better name for `SExpr0`, I think  it is a good idea to rename those in ANF.scala (here source/target make complelty sens) and in the Compiler.scala (at least target makes sens here) because otherwise it is completely unreadable, but please:
- do not rename it in other files: I think other files should simply not use SExpr0 at all (we can take care of that in another PR) 
- use `source.` and `target.` instead of `s.` and `t.`: I am fine to use small cap letter for local variables only.

**Review by cocreature — APPROVED** [review]
Nice work, thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 5]
```diff
@@ -448,8 +393,9 @@ object SExpr {
     def ref: DefinitionRef
     def packageId: PackageId = ref.packageId
     def modName: ModuleName = ref.qualifiedName.module
-    private[this] val eval = SEVal(this)
-    def apply(args: SExpr*) = SEApp(eval, args.toArray)
+    private[this] val eval = s.SEVal(this)
+    def apply(args: s.SExpr*) = s.SEApp(eval, args.toArray) // build at compile-time
+    def apply(args: SExpr*) = SEApp(SEVal(this), args.toArray) // build at runtime
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:398]: And two obsious changes in `Compiler` and `Anf`

**Comment by nickchapman-da** [comment]
Thank you Remy and Moritz for your feedback. Many are addressed in this PR. Also, I added some TODO markers where Remy suggests we might move code from the _Expression_ classes and into the _Compiler_.

One small thing I want to push back on: The change where the `SExpr` type (abstract trait) is moved into the `SExpr` object. And the same pattern in the `SExpr0` object.

Remy said "If want SEXpr accessible in companion object use an alias". But the point is that I wanted *not* to have access to the `SExpr` type outside the object. i.e. I wanted the `SExpr`/`SExpr0` objects as single roots for all the types defined in their respective files. Having additional `SExpr` and `SExpr0` types floating about at the top level seemed an unnecessary and inconvenient special case.

Perhaps there is a different name-spacing mechanism in scala (packages?) which would serve this purpose better, but I wanted the code abuse done in this PR to remain focused on the task at hand. :)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 6]
```diff
@@ -112,6 +104,8 @@ object SExpr {
 
   /** Function application:
     *    General case: 'fun' and 'args' are any kind of expression
+    *    This case *ought* not to exist post-ANF.
+    *    Sadly, we have many code paths where this case is constructed.
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:108]: There are various code-paths in `SBuiltin` where we construct _non-atomic_ applications at runtime. `SEApp(SEVal(...))` For example:
```
SEApp(SEVal(KeyDefRef(templateId)), Array(SELocS(1)))
```
```
SEApp(SEVal(SignatoriesDefRef(templateId)), Array(SELocS(1))),
```
We also have places in our test code where these applications are constructed.

I totally do want to remove `SEAppAtomicGeneral` from the expression forms which run on the machine. I also want us to stop manufacturing expressions at runtime. I will start an issue for tracking these and other refactoring goals.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala [hunk 7]
```diff
@@ -183,90 +187,113 @@ class AnfTest extends AnyWordSpec with Matchers {
...
-  private def testTransform(original: SExpr, expected: SExpr, show: Boolean = false): Assertion = {
+  private def testTransform(
+      original: s.SExpr,
+      expected: t.SExpr,
+      show: Boolean = false,
+  ): Assertion = {
     val transformed = flattenToAnf(original)
     if (show || transformed != expected) {
-      println(s"**original:\n${pp(original)}\n")
+      //println(s"**original:\n${pp(original)}\n")
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala]: I didn't want to duplicate the pretty print code for the `SExpr0` types, just for this one place.
I reinstated the test-fail print just the vanilla `toString`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  * 1: convert from LF
+  *     - reducing binding forms (update & scenario becoming builtins)
+  *     - moving to de Bruijn indexing for variable
+  *     - moving to multi-argument applications and abstractions.
+  *
+  * 2: closure conversion
+  * 3: transform to ANF
+  *
+  * Stages 1 and 2 are in Compiler.scala; stage 3 in Anf.scala.
+  * During Stage3 (ANF transformation), we move from this type (SExpr0) to SExpr,
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala:23]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala]: 1. Pushing back on this. Please see my free-standing comment at the bottom of the PR.
2. I added private[speedy] to `SExpr0`. Thank you for spotting this oversight

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: I investigated this _sharing loss_ some more... I inlined the single use of the _runtime_ `apply` method, into it's caller in `SBCallInterface.execute`. Please see commit: 5eb725688e.

The code is now:
```
    override private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
      val record = getSRecord(args, 0)
      machine.ctrl =
        SEApp(SEVal(ImplementsMethodDefRef(record.id, ifaceId, methodName)), Array(SEValue(record)))
    }
```
I think this shows that there never was any useful sharing, and I don't think there can be. The argument of `SEVal`, `ImplementsMethodDefRef(...)` is constructed dynamically from dynamic data.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 9]
```diff
@@ -19,24 +24,26 @@ import com.daml.lf.speedy.SValue._
...
 @SuppressWarnings(Array("org.wartremover.warts.Any"))
 object SExpr {
 
+  sealed abstract class SExpr extends Product with Serializable {
+    def execute(machine: Machine): Unit
+    @SuppressWarnings(Array("org.wartremover.warts.Any"))
+    override def toString: String =
+      productPrefix + productIterator.map(prettyPrint).mkString("(", ",", ")")
+  }
+
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:46]: see comment at end

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 10]
```diff
@@ -47,8 +49,11 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
 
   // Helper for constructing expressions applying this builtin.
   // E.g. SBCons(SEVar(1), SEVar(2))
-  private[lf] def apply(args: SExpr*): SExpr =
-    SEApp(SEBuiltin(this), args.toArray)
+  private[lf] def apply(args: s.SExpr*): s.SExpr = //expressions built at compile-time
+    s.SEApp(s.SEBuiltin(this), args.toArray)
+
+  private[lf] def apply(args: t.SExpr*): t.SExpr = //expressions built at runtime
+    t.SEApp(t.SEBuiltin(this), args.toArray)
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: I added `TODO` markers in various places where you suggest code which should/could be moved into the compiler.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: I stopped using the renames `source`/`target`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 11]
```diff
@@ -19,24 +24,26 @@ import com.daml.lf.speedy.SValue._
 import com.daml.lf.speedy.Speedy._
 import com.daml.lf.speedy.SError._
 import com.daml.lf.speedy.SBuiltin._
-import com.daml.nameof.NameOf
+import com.daml.lf.speedy.{SExpr0 => s}
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: I avoided the unhelpful `s.` prefix

**Comment by remyhaemmerle-da** [comment]
@nickchapman-da OK. I am fine you use your trick inside speedy. 
Ideally I would have like that this is not visible outside, meaning we do not have to use `SExpr.SExpr` in the engine, script and triggers. That say, as speedy is already leaking so much it is probably not so important. I except somebody touching this component those components should be familiar with Speedy. 

However, you ignored the remarks about import renaming.

**Comment by remyhaemmerle-da** [comment]
In addition to what I write in the review, I think it is a good idea to roughly follow [Scala Name Convention](https://docs.scala-lang.org/style/naming-conventions.html) in order to respect the principle of least surprise.

**Pull request merged**